pkg/progressreader/progressstatus.go

```diff
@@ -0,0 +1,78 @@
+package progressreader
+
+import (
+	"bytes"
+	"io"
+	"sync"
+)
+
+// ProgressStatus keeps track of one or more observers watching the progress
+// of an operation. For example, if multiple clients are trying to pull an
+// image, they share a ProgressStatus for the download operation.
+type ProgressStatus struct {
+	sync.Mutex
+	// c is a channel that observers block on, waiting for the operation
+	// to finish.
+	c chan struct{}
+	// observers is a list of io.Writers which receive a copy of the
+	// progress output.
+	observers []io.Writer
+	// history is a buffer of the progress output so far, so a new observer
+	// can catch up.
+	history bytes.Buffer
+}
+
+// NewProgressStatus returns a ProgressStatus structure
+func NewProgressStatus() *ProgressStatus {
+	return &ProgressStatus{
+		c:         make(chan struct{}),
+		observers: []io.Writer{},
+	}
+}
+
+// Write adds data to the history buffer, and also writes it to all current
+// observers.
+func (ps *ProgressStatus) Write(p []byte) (n int, err error) {
+	ps.Lock()
+	defer ps.Unlock()
+	ps.history.Write(p)
+	for _, w := range ps.observers {
+		// copy paste from MultiWriter, replaced return with continue
+		n, err = w.Write(p)
+		if err != nil {
+			continue
+		}
+		if n != len(p) {
+			err = io.ErrShortWrite
+			continue
+		}
+	}
+	return len(p), nil
+}
+
+// AddObserver adds an observer to the ProgressStatus. The new observer
+// receives the data from the history buffer, and also all subsequent data.
+func (ps *ProgressStatus) AddObserver(w io.Writer) {
+	ps.Lock()
+	defer ps.Unlock()
+	w.Write(ps.history.Bytes())
+	ps.observers = append(ps.observers, w)
+}
+
+// Done signals to all observers that the operation has finished.
+func (ps *ProgressStatus) Done() {
+	ps.Lock()
+	close(ps.c)
+	ps.history.Reset()
+	ps.Unlock()
+}
+
+// Wait blocks until the operation is marked as completed by the Done method.
+func (ps *ProgressStatus) Wait() error {
+	ps.Lock()
```

Comment:
This can simply be the following:

```
<-ps.c
return nil
```

There is no need to lock it. The field isn't (and shouldn't) be written concurrently.